张量：1.多维数组 2.variable(已经被并入tensor) 用于自动求导

dtype:数据类型 torch.long torch.float
shape:形状
device:张量所在设备
requires_grad
grad
grad_fn
is_leaf

torch.tensor()
功能：从data创建tensor
data：数据，可以是list，numpy
dtype：数据类型，默认与data的一到device：所在设备，cuda/cpu
requires_grad：是否需要梯度
pin_memory：是否存于锁页内存 通常为false

### 直接创建

In [5]:
#通过np.array创建
import torch
import numpy as np
#直接创建
arr = np.ones((3,3))
t = torch.tensor(arr)
print(t)

tensor([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]], dtype=torch.float64)


torch.from_numpy(ndarray)
功能：从numpy创建tensor
注意事项：从torch.from_numpy创建的tensor与原ndarray共享内存，当修改其中一个的数据，另外一个也将会被改动

In [7]:
t = torch.from_numpy(arr)
print(t)
arr[0,0] = 0
print(t)
t[0,0] = -1
print(arr)

tensor([[0., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]], dtype=torch.float64)
tensor([[0., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]], dtype=torch.float64)
[[-1.  1.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]]


### 根据数值创建

2.1 torch.zeros()
功能：依size创建全0张量
size：张量的形状，如（3，3）、（3，224，224） out：输出的张量
layout：内存中布局形式，有strided（默认），sparse_coo（稀疏矩阵的时候）等
device：所在设备，gpu/cpu
requires_grad ：是否需要梯度

In [9]:
out_t = torch.tensor([1])
t = torch.zeros((3,3), out = out_t)
print(t,'\n',out_t)
# 内存地址也是一样的
print(id(t),id(out_t))

tensor([[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]]) 
 tensor([[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]])
140586769394432 140586769394432


2.2 torch.zeros_like()
功能：依input形状创建全0张量
intput：创建与input同形状的全0张量 dtype 数据类型
layout：内存中布局形式
2.3 torch.ones()
2.4 torch.ones_like()
功能：依input形状创建全1张量
2.5 torch.full()
2.6 torch.full_like()
功能：依input形状创建全0张量 size：张量的形状，如（3，3） fill_value：张量的值

In [10]:
t = torch.full((3,3),8)
t

tensor([[8, 8, 8],
        [8, 8, 8],
        [8, 8, 8]])

2.7 torch.arange()
功能：创建等差的1维张量
注意事项：数值区间为【start，end）start：数列起始值
end：数列“结束值"
step：数列公差，默认为1

In [13]:
t = torch.arange(2,10,2)
t

tensor([2, 4, 6, 8])

2.8 torch.linspace()
功能：创建均分的1维张量
注意事项：数值区间为【start，end】start：数列起始值
end：数列结束值
steps：数列长度

In [16]:
t = torch.linspace(2,10,6)
t

tensor([ 2.0000,  3.6000,  5.2000,  6.8000,  8.4000, 10.0000])

2.9 torch.logspace()
功能：创建对数均分的1维张量
注意事项：长度为steps，底为base
start：数列起始值
end：数列结束值
steps：数列长度
base：对数函数的底，默认为10
2.10 torch.eye()
功能：创建单位对角矩阵（2维张量）注意事项：默认为方阵
n：矩阵行数
m ：矩阵列数

In [18]:
t = torch.logspace(1,10,5)
t

tensor([1.0000e+01, 1.7783e+03, 3.1623e+05, 5.6234e+07, 1.0000e+10])

In [20]:
t = torch.eye(3,6)
t

tensor([[1., 0., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0., 0.],
        [0., 0., 1., 0., 0., 0.]])

### 依概率分布创建张量

3.1 torch.normal()
功能：生成正态分布（高斯分布）
mean：均值
std：标准差

四种模式：
mean为标量，std为标量
mean为标量，std为张量
mean为张量，std为标量
mean为张量，std为张量

In [23]:
#mean标量 std标量
t = torch.normal(0.,1.,size=(4,))
t

tensor([ 0.4520, -0.0375,  1.3191, -1.0902])

In [25]:
#mean张量 std标量
mean = torch.arange(1,5,dtype = float)
std = 1
t = torch.normal(mean,std)
t

tensor([-0.0701,  2.0102,  2.0279,  2.8081], dtype=torch.float64)

3.2 torch.randn()
3.3 torch.randn like() 功能：生成标准正态分布
size：张量的形状
3.4 torch.rand()
3.5 torch.rand like()
功能：在区间【0，1）上，生成均匀分布
3.6 torch.randint()
3.7 torch.randint_like()
功能：区间【low，high）生成整数均匀分布size：张量的形状
3.8 torch.randperm()
功能：生成生成从0到n-1的随机排列n：张量的长度
用来生成乱序索引
3.9 torch.bernoulli()
功能：以input为概率，生成伯努利分布（0-1分布，两点分布）
input：概率值